File: funnyvote/app/src/test/java/com/heaton/funnyvote/data/promotion/PromotionRepositoryTest.java
package com.heaton.funnyvote.data.promotion;

import com.heaton.funnyvote.database.Promotion;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import rx.observers.TestSubscriber;

import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class PromotionRepositoryTest {

    @Mock
    private PromotionDataSource localPromotionDataSource;
    @Mock
    private PromotionDataSource remotePromotionDataSource;

    private PromotionRepository promotionRepository;

    private static User user = new User();
    private static List<Promotion> promotionList = new ArrayList<>();

    @Before
    public void setUpPromotionRepository() {
        MockitoAnnotations.initMocks(this);
        promotionRepository = PromotionRepository.getInstance(remotePromotionDataSource, localPromotionDataSource);
    }

    @After
    public void destroyRepositoryInstance() {
        PromotionRepository.destroyInstance();
    }

    @Test
    public void getPromotionList_getFromRemoteAndSaveToLocal() {
        when(remotePromotionDataSource.getPromotionList(eq(user))).thenReturn(rx.Observable.just(promotionList));
        when(localPromotionDataSource.getPromotionList(eq(user))).thenReturn(rx.Observable.error(new Exception()));
        when(promotionRepository.getPromotionList(user)).thenReturn(rx.Observable.just(promotionList));

        TestSubscriber<List<Promotion>> testSubscriber1 = new TestSubscriber<>();
        TestSubscriber<List<Promotion>> testSubscriber2 = new TestSubscriber<>();


        promotionRepository.getPromotionList(user).subscribe(testSubscriber1);
        verify(remotePromotionDataSource).getPromotionList(eq(user));
        //getPromotionsCallbackArgumentCaptor.getValue().onPromotionsLoaded(anyList());
        //verify(localPromotionDataSource, times(2)).savePromotionList(promotionList);

    }

    @Test
    public void getPromotionListWithRemoteFailure_getFromLocal() {
        when(localPromotionDataSource.getPromotionList(eq(user))).thenReturn(rx.Observable.just(promotionList));
        when(remotePromotionDataSource.getPromotionList(user)).thenReturn(rx.Observable.error(new Exception()));
        when(promotionRepository.getPromotionList(user)).thenReturn(rx.Observable.just(promotionList));
        promotionRepository.getPromotionList(user);
        verify(remotePromotionDataSource).getPromotionList(eq(user));
        //getPromotionsCallbackArgumentCaptor.getValue().onPromotionsNotAvailable();
        verify(localPromotionDataSource, never()).savePromotionList(eq(promotionList));
        verify(localPromotionDataSource, times(2)).getPromotionList(eq(user));
        //getPromotionsCallbackArgumentCaptor.getValue().onPromotionsLoaded(promotionList);
        //verify(getPromotionsCallback).onPromotionsLoaded(promotionList);
    }
}

File: funnyvote/app/src/test/java/com/heaton/funnyvote/data/promotion/PromotionRepositoryTest.kt
package com.heaton.funnyvote.data.promotion

import com.heaton.funnyvote.capture
import com.heaton.funnyvote.database.Promotion
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.eq
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.*
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import java.util.*

class PromotionRepositoryTest {

    @Mock
    private lateinit var localPromotionDataSource: PromotionDataSource
    @Mock
    private lateinit var remotePromotionDataSource: PromotionDataSource

    private lateinit var promotionRepository: PromotionRepository
    @Mock
    private lateinit var getPromotionsCallback: PromotionDataSource.GetPromotionsCallback

    @Captor
    private lateinit var getPromotionsCallbackArgumentCaptor: ArgumentCaptor<PromotionDataSource.GetPromotionsCallback>

    @Before
    fun setUpPromotionRepository() {
        MockitoAnnotations.initMocks(this)
        promotionRepository = PromotionRepository.getInstance(remotePromotionDataSource, localPromotionDataSource)
    }

    @After
    fun destroyRepositoryInstance() {
        PromotionRepository.destroyInstance()
    }

    @Test
    fun getPromotionList_getFromRemoteAndSaveToLocal() {
        promotionRepository.getPromotionList(user, getPromotionsCallback)
        verify<PromotionDataSource>(remotePromotionDataSource).getPromotionList(eq(user), capture(getPromotionsCallbackArgumentCaptor))
        getPromotionsCallbackArgumentCaptor.value.onPromotionsLoaded(ArgumentMatchers.anyList())
        verify<PromotionDataSource>(localPromotionDataSource).savePromotionList(promotionList)

    }

    @Test
    fun getPromotionListWithRemoteFailure_getFromLocal() {
        promotionRepository.getPromotionList(user, getPromotionsCallback)
        verify<PromotionDataSource>(remotePromotionDataSource).getPromotionList(eq(user), capture(getPromotionsCallbackArgumentCaptor))
        getPromotionsCallbackArgumentCaptor.value.onPromotionsNotAvailable()
        verify<PromotionDataSource>(localPromotionDataSource, never())
                .savePromotionList(eq(promotionList))
        verify<PromotionDataSource>(localPromotionDataSource).getPromotionList(eq(user), capture(getPromotionsCallbackArgumentCaptor))
        getPromotionsCallbackArgumentCaptor.value.onPromotionsLoaded(promotionList)
        verify<PromotionDataSource.GetPromotionsCallback>(getPromotionsCallback).onPromotionsLoaded(promotionList)
    }

    companion object {

        private val user = User()
        private val promotionList = ArrayList<Promotion>()
    }
}
